src/neural/xla/onnx2hlo.cc

```diff
@@ -458,13 +587,87 @@ class Onnx2HloConverter {
                             is_sorted, is_unique)};
   }
 
+  HloFlow DoReduceMean(HloFlow input, const std::vector<int64_t>& axes) {
+    HloFlow zero = MakeScalar(0, input->shape().element_type());
+    auto flow = builder_.Reduce(
+        input, zero, MakeAddComputation(input->shape().element_type()), axes);
+    size_t count = std::accumulate(
+        axes.begin(), axes.end(), 1, [&](size_t acc, size_t axis) {
+          return acc * input->shape().dimensions(axis);
+        });
+    auto denominator =
+        DoBroadcast(MakeScalar(count, input->shape().element_type()),
+                    flow->shape().dimensions());
+    flow = builder_.Divide(flow, denominator);
+    HloTensorType target_shape(input->shape());
+    for (auto axis : axes) target_shape.SetDimension(axis, 1);
+    return builder_.Reshape(flow, target_shape);
+  }
+
+  std::vector<HloFlow> OpReduceMean(const pblczero::NodeProto& node) {
+    CheckKnownAttributes(node, 1, {"axes"});
+    auto* input = GetInput(node, 0);
+    auto axes = GetAttribute(node, "axes")->ints();
+    return {DoReduceMean(input, axes)};
+  }
+
+  std::vector<HloFlow> OpCast(const pblczero::NodeProto& node) {
+    CheckKnownAttributes(node, 1, {"to"});
+    auto* input = GetInput(node, 0);
+    const auto onnx_type = static_cast<pblczero::TensorProto::DataType>(
+        GetAttribute(node, "to")->i());
+    const auto hlo_type = OnnxTypeToXlaType(onnx_type);
+    if (input->shape().element_type() == hlo_type) return {input};
+    return {builder_.Convert(input, hlo_type)};
+  }
+
+  std::vector<HloFlow> OpLayerNormalization(const pblczero::NodeProto& node) {
+    CheckKnownAttributes(node, 3, {"axis", "epsilon"});
+    auto* input = GetInput(node, 0);
+    const auto axis = GetAttribute(node, "axis")->i();
+    const auto epsilon = GetAttribute(node, "epsilon")->f();
+    auto* scale = GetInput(node, 1);
+    auto* bias = GetInput(node, 2, true);
+    const bool need_conv =
+        input->shape().element_type() != pblczero::XlaShapeProto::F32;
+    auto* flow = need_conv
+                     ? builder_.Convert(input, pblczero::XlaShapeProto::F32)
+                     : input;
+    flow = DoBroadcast(DoReduceMean(flow, {axis}), input->shape().dimensions());
+    auto* norm = builder_.Subtract(input, flow);
+    flow = builder_.Multiply(norm, norm);
+    flow = DoReduceMean(flow, {axis});
+    flow = builder_.Add(
+        flow, DoBroadcast(MakeScalar(epsilon, pblczero::XlaShapeProto::F32),
```

Comment:
Indeed elementwise ops support broadcast (didn't know)!  
They don't do that implicitly, but accept same `broadcast_dimensions` attribute as the broadcast op. That will simplify the code, I'll do that today or tomorrow.

Reply:
Consider the approval valid even if you similarly change other parts of the code.

Reply:
I'll merge the PR as it is, will simplify binary ops in a different PR.

Reply:
I've just tried this.  
Turns out that although Xla's `XlaBuilder` allows implicit broadcasting while inserting binary ops, it actually inserts explicit broadcast instructions to the final HLO.

Initially I thought one `HloBuilder` call = one HLO instruction is a good idea, but if `XlaBuilder` inserts several, maybe we can do too, but I'm not sure.  
It would carry some complexity from onnx2hlo into hlo_builder (https://github.com/LeelaChessZero/lc0/compare/master...mooskagh:lc0:xla-8), but I don't think that the current onnx2hlo is too bad and not sure whether that move could be considered an improvement.

Reply:
Doesn't seem worth the trouble.